vm/systemSmartContracts/governance.go

```diff
@@ -288,521 +499,599 @@ func (g *governanceContract) whiteListProposal(args *vmcommon.ContractCallInput)
+	return nil
+}
+
+// getGeneralProposal returns a proposal from storage
+func (g *governanceContract) getGeneralProposal(reference []byte) (*GeneralProposal, error) {
+	key := append([]byte(proposalPrefix), reference...)
+	marshaledData := g.eei.GetStorage(key)
+
+	if len(marshaledData) == 0 {
+		return nil, vm.ErrProposalNotFound
 	}
-	err = g.saveGeneralProposal(gitHubCommit, generalProposal)
+
+	generalProposal := &GeneralProposal{}
+	err := g.marshalizer.Unmarshal(generalProposal, marshaledData)
 	if err != nil {
-		log.Warn("saveGeneralProposal", "err", err)
-		g.eei.AddReturnMessage("saveGeneralProposal" + err.Error())
-		return vmcommon.UserError
+		return nil, err
 	}
 
-	return vmcommon.Ok
+	return generalProposal, nil
 }
 
-func (g *governanceContract) vote(args *vmcommon.ContractCallInput) vmcommon.ReturnCode {
-	if args.CallValue.Cmp(zero) != 0 {
-		g.eei.AddReturnMessage("invalid proposal cost, expected 0")
-		return vmcommon.OutOfFunds
-	}
-	err := g.eei.UseGas(g.gasCost.MetaChainSystemSCsCost.Vote)
+// proposalExists returns true if a proposal already exists
+func (g *governanceContract) proposalExists(reference []byte) bool {
+	key := append([]byte(proposalPrefix), reference...)
+	marshaledData := g.eei.GetStorage(key)
+	return len(marshaledData) > 0
+}
+
+// getValidProposal returns a proposal from storage if it exists or it is still valid/in-progress
+func (g *governanceContract) getValidProposal(reference []byte) (*GeneralProposal, error) {
+	proposal, err := g.getGeneralProposal(reference)
 	if err != nil {
-		g.eei.AddReturnMessage("not enough gas")
-		return vmcommon.OutOfGas
+		return nil, err
 	}
-	if len(args.Arguments) < 2 || len(args.Arguments) > 3 {
-		g.eei.AddReturnMessage("invalid number of argument expected 2 or 3")
-		return vmcommon.FunctionWrongSignature
+
+	currentNonce := g.eei.BlockChainHook().CurrentNonce()
+	if currentNonce < proposal.StartVoteNonce {
+		return nil, vm.ErrVotingNotStartedForProposal
 	}
-	if len(args.Arguments) == 3 && len(args.Arguments[2]) != len(args.CallerAddr) {
-		g.eei.AddReturnMessage("wrong argument number 3 should be a valid address")
-		return vmcommon.FunctionWrongSignature
+
+	if currentNonce > proposal.EndVoteNonce {
+		return nil, vm.ErrVotedForAnExpiredProposal
 	}
-	if len(args.Arguments) == 3 && bytes.Equal(args.CallerAddr, args.Arguments[2]) {
-		g.eei.AddReturnMessage("wrong argument number 3 should be different than caller")
-		return vmcommon.FunctionWrongSignature
+
+	return proposal, nil
+}
+
+// isWhiteListed checks if an address is whitelisted
+func (g *governanceContract) isWhiteListed(address []byte) bool {
+	key := append([]byte(whiteListPrefix), address...)
+	marshaledData := g.eei.GetStorage(key)
+	if len(marshaledData) == 0 {
+		return false
 	}
 
-	proposalToVote := args.Arguments[0]
-	if !g.proposalExists(proposalToVote) {
-		g.eei.AddReturnMessage("proposal does not exists")
-		return vmcommon.UserError
+	key = append([]byte(proposalPrefix), address...)
```

Comment:
Not sure I understand the question, but this is from the old implementation and this function first checks if there is an entry for the address to be whitelisted, then also checks if there actually exists a voted proposal for it.
Q for @sasurobert though, because now I realise that `whiteListAtGenesis` was not actually called anywhere, so we can remove that code, right? And save whitelist proposals in initV2 for the `InitalWhiteListedAddresses` I added?

Reply:
Yeah, it seems ok after I read the whole code. Still, from a workflow perspective: before I can add a new proposal, I have to first vote to become a whitelisted address, right? 
Also, we can use whitelist at genesis to enable some addresses to become automatically whitelisted when we activate the feature.

Reply:
if whiteListAtGenesis is not used, than it has to be implemented in systemSCs.go with enable flag, calling and setting the first whitelisted address.

Reply:
Yup, I think we need that before trying to test the voting in a system test. Might have another PR for that